### PR TITLE
Backport #453 to Java branch

### PR DESCRIPTION
**Pull request by cmacdonald**
This is my attempt to backport #453 to the Java branch.

Some comments:
 - 87ccb05f120c121194eff647aa275bd886483138 makes it easier to set the terrier and helper versions
 - bf785eea6bccdb5cb61cd0e787b6362207c09023 fixes the `pt.init()` replacement warning message for `pt.java.add_package()` [which no longer appears in test_rewrite_rm3.py anyway]
 - @seanmacavaney I think we agreed that we would have something like `pt.rewrite.enable_extra_prf()` for AxiomaticQE etc. One extra jar file for every pyterrier start is an unnecessary delay [but Axiomatic QE is the only thing that would be left once RM3 is moved to terrier-assemblies jar file]
 - The test_rewrite_rm3.py doesnt pass for me on this branch?

```
E       AssertionError: 'appl[28 chars]367 sideband^0.027371584 ferrit^0.027371584 mo[122 chars]6367' != 'appl[28 chars]367 suppli^0.037346367 see^0.027371584 design^[119 chars]1584'
E       - applypipeline:off equip^0.037346367 sideband^0.027371584 ferrit^0.027371584 modul^0.027371584 suppli^0.037346367 design^0.056739070 unit^0.037346367 anod^0.037346367 compact^0.674414337 stabil^0.037346367
E       + applypipeline:off equip^0.037346367 suppli^0.037346367 see^0.027371584 design^0.056739070 microwav^0.027371584 anod^0.037346367 unit^0.037346367 compact^0.674414337 stabil^0.037346367 singl^0.027371584

```

### CODE REVIEW

File: pyterrier/terrier/java.py [hunk 1]
```diff
@@ -23,17 +23,20 @@
 
 @pt.java.before_init
 def set_version(version: Optional[str] = None):
-    configure['terrier_version'] = version
```
Comment (seanmacavaney) [inline pyterrier/terrier/java.py:26]: I don't understand this logic -- what does this simplify? There's also no way to reset the terrier_version to default (latest version) anymore?

**Review by seanmacavaney — COMMENTED** [review]
> I think we agreed that we would have something like pt.rewrite.enable_extra_prf() for AxiomaticQE etc. One extra jar file for every pyterrier start is an unnecessary delay [but Axiomatic QE is the only thing that would be left once RM3 is moved to terrier-assemblies jar file]

Whoops, I thought we decided to move it all into terrier-assemblies eventually, and that we didn't want the friction of enable_prf at all. My bad.

Any reason not to just move AQE into terrier-assemblies? It seems silly to have an assembly, enable_extra_prf() etc. just for one model.

**Comment by cmacdonald** [comment]
> Any reason not to just move AQE into terrier-assemblies? It seems silly to have an assembly, enable_extra_prf() etc. just for one model.

I dont think its a full reproduction. I couldnt get the performances to match Jimmy's paper. So in this way, it was experimental.

File: .github/workflows/push.yml [hunk 2]
```diff
@@ -79,6 +79,13 @@ jobs:
         # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
         #flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
 
+    - name: RM3 unit tests
```
[the remark this answers is not in the document]
Comment (cmacdonald) [reply .github/workflows/push.yml:82]: awaiting decision on terrier-prf etc.

File: pyterrier/java/_utils.py [hunk 3]
```diff
@@ -209,8 +209,10 @@ def legacy_init(version=None, mem=None, packages=[], jvm_opts=[], redirect_io=Tr
         deprecated_calls.append(f'pt.java.set_log_level({logging!r})')
 
     for package in boot_packages:
-        pt.java.add_package(*package.split(':')) # format: org:package:version:filetype (where version and filetype are optional)
-        deprecated_calls.append(f'pt.java.add_package({package!r})')
+        pkg_split = package.split(':')
+        pkg_string = ", ".join(f'"{w}"' for w in pkg_split)
```
[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/java/_utils.py:213]: I had nightmares with fstrings inside fstrings. let me try again.

**Comment by seanmacavaney** [comment]
Maybe we should drop it entirely then? Or move it out of the core repo into an extension package?

File: pyterrier/java/_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (seanmacavaney) [reply pyterrier/java/_utils.py]:
```
f'{w!r}' for w in pkg_split
```
or
```
repr(w) for w in pkg_split
```
should do the trick!

[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/java/_utils.py]: its fixed

File: pyterrier/terrier/java.py (hunk 1 above)
Comment (seanmacavaney) [reply pyterrier/terrier/java.py:26]: I think we should have a different sentinal value for "use latest version" then.

Comment (cmacdonald) [reply pyterrier/terrier/java.py:26]: None is the latest version, right?

Comment (seanmacavaney) [reply pyterrier/terrier/java.py:26]: yeah

**Comment by mam10eks** [comment]
Interesting, good catch!
As far as I see, the problem are score ties, or?

I.e., on one machine, the term `modul` is included with a score `0.027371584`, and on another machine the term `singl` is included with the same score of `0.027371584`.

So we basically could solve this in two directions, one could be to relax the tests (e.g., only testing against the sorted score vector and maybe the inclusion of the top-5 terms with disjoint terms that should be deterministic). Another idea that I would also find very interesting is to break the score ties during expansion in a deterministic way, this should also resolve the problem.

What do you think about the possible options? I would be happy to help :)

**Comment by cmacdonald** [comment]
Great observation, its a tie break problem. The relevant sort code is here:
https://github.com/terrierteam/terrier-prf/blob/master/src/main/java/org/terrier/querying/RM1.java#L251-L266

I think we can place a tie-breaker. Terrier, trec_eval etc all have tie-breakers for docnos. Breaking ties on termids would be appropriate, and those would be stable unless the underlying index changed (the manner that termids are assigned is different from classical vs single-pass indexing).

**Comment by mam10eks** [comment]
We could also think about using the md5 of the termid, as this would make it "pseudo random"?

I could prepare a proposal in the Java code above?

**Comment by cmacdonald** [comment]
I've made the tie-break change at:
https://github.com/terrierteam/terrier-prf/commit/ad922751cff654bb9435b0adaafdd6d84b3e5576?diff=unified&w=1
This fixes the tie-break problem.

I had to update the unit test https://github.com/terrier-org/pyterrier/pull/454/commits/4eb0799ae7e327546e4adfdcc6157f70d5e85591 based on weights obtained on my mac (My intel mac is on Java 17).

Unfortunately, these dont match what the Ubuntu gives in Github Actions!
`AssertionError: 'appl[25 chars]264644 ferrit^0.025508024 modul^0.025508024 su[125 chars]4644' != 'appl[25 chars]264640 ferrit^0.025508020 modul^0.025508020 su[125 chars]4640'`

 Clearly such arithmetic noise -- changed at the 9th decimal place -- are pretty unlikely to have an impact on ranking, so we just need a way to make the unit test a little more robust to these (platform specific? java version specific?) rounding errors. Regex?

**Comment by mam10eks** [comment]
Cool, yes, making the test more robust makes sense! I am personally not a fan of regular expressions, so I would personally prefer a method like:

```
def normalize_term_weights(term_weights, digits=7):
    ret = ''
    for i in term_weights.split():
        if '^' in i:
            i = i.split('^')
            i = i[0] + '^' + i[1][:digits]
        ret += ' ' + i
    return ret.strip()
```

This is maybe not the most elegant solution, but might suffice :)
